File: day10/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

type (
	Point struct {
		Posx int
		Posy int
		Velx int
		Vely int
	}

	PointList []*Point
)

func (p *Point) Step(t int) {
	p.Posx += p.Velx * t
	p.Posy += p.Vely * t
}

func (p PointList) Step(t int) {
	for _, i := range p {
		i.Step(t)
	}
}

func (p PointList) Size() (int, int, int, int) {
	minx := 99999999
	maxx := -99999999
	miny := 99999999
	maxy := -99999999
	for _, i := range p {
		if i.Posx < minx {
			minx = i.Posx
		}
		if i.Posx > maxx {
			maxx = i.Posx
		}
		if i.Posy < miny {
			miny = i.Posy
		}
		if i.Posy > maxy {
			maxy = i.Posy
		}
	}
	sizex := maxx - minx + 1
	sizey := maxy - miny + 1
	return sizex, sizey, minx, miny
}

func (p PointList) Print() {
	sizex, sizey, minx, miny := p.Size()

	baseline := make([]byte, 0, sizex)
	for i := 0; i < sizex; i++ {
		baseline = append(baseline, '.')
	}
	board := make([][]byte, sizey)
	for i := range board {
		board[i] = make([]byte, sizex)
		copy(board[i], baseline)
	}

	for _, i := range p {
		x := i.Posx - minx
		y := i.Posy - miny
		board[y][x] = '#'
	}

	for _, i := range board {
		fmt.Println(string(i))
	}
}

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	points := PointList{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		var posx, posy, velx, vely int
		fmt.Sscanf(scanner.Text(), "position=<%d, %d> velocity=<%d, %d>", &posx, &posy, &velx, &vely)
		points = append(points, &Point{
			Posx: posx,
			Posy: posy,
			Velx: velx,
			Vely: vely,
		})
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	step := 10656
	points.Step(step)
	points.Print()
	fmt.Println("Step: ", step)
	sizex, sizey, minx, miny := points.Size()
	fmt.Println(sizex, sizey, minx, miny)
}

File: day10/src/main.rs
use regex::Regex;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

struct PointList {
    points: Vec<(isize, isize, isize, isize)>,
}

impl PointList {
    fn new() -> Self {
        Self { points: Vec::new() }
    }

    fn add(&mut self, point: (isize, isize, isize, isize)) {
        self.points.push(point);
    }

    fn step(&mut self, t: isize) {
        for &mut (ref mut x, ref mut y, vx, vy) in self.points.iter_mut() {
            *x += t * vx;
            *y += t * vy;
        }
    }

    fn size(&self) -> (usize, usize, isize, isize) {
        let (mut minx, mut miny, _, _) = self.points[0];
        let (mut maxx, mut maxy, _, _) = self.points[0];
        for &(x, y, _, _) in self.points.iter().skip(1) {
            if x < minx {
                minx = x;
            }
            if x > maxx {
                maxx = x;
            }
            if y < miny {
                miny = y;
            }
            if y > maxy {
                maxy = y;
            }
        }
        (
            (maxx - minx + 1) as usize,
            (maxy - miny + 1) as usize,
            minx,
            miny,
        )
    }

    fn print(&self) {
        let (sizex, sizey, minx, miny) = self.size();
        let mut board = vec![vec!['.'; sizex]; sizey];
        for &(x, y, _, _) in self.points.iter() {
            let x = (x - minx) as usize;
            let y = (y - miny) as usize;
            board[y][x] = '#';
        }
        for i in board {
            println!("{}", i.iter().collect::<String>());
        }
        println!("{} {} {} {}", sizex, sizey, minx, miny);
    }
}

fn main() {
    let re = Regex::new(
        r"^.*< ?(?P<posx>-?\d+),  ?(?P<posy>-?\d+)>.*< ?(?P<velx>-?\d+),  ?(?P<vely>-?\d+)>$",
    )
    .expect("Invalid regex");
    let file = File::open(PUZZLEINPUT).expect("Failed to open file");
    let reader = BufReader::new(file);

    let mut points = PointList::new();
    for line in reader.lines() {
        let l = line.expect("Failed to read line");
        let caps = re.captures(&l).expect("Regex cannot capture line");
        let posx = caps
            .name("posx")
            .expect("Posx does not exist")
            .as_str()
            .parse::<isize>()
            .expect("Failed to parse posx");
        let posy = caps
            .name("posy")
            .expect("Posy does not exist")
            .as_str()
            .parse::<isize>()
            .expect("Failed to parse posy");
        let velx = caps
            .name("velx")
            .expect("Velx does not exist")
            .as_str()
            .parse::<isize>()
            .expect("Failed to parse velx");
        let vely = caps
            .name("vely")
            .expect("Vely does not exist")
            .as_str()
            .parse::<isize>()
            .expect("Failed to parse vely");
        points.add((posx, posy, velx, vely));
    }

    let step = 10656;
    points.step(step);
    points.print();
    println!("Step: {}", step);
}
